proposals/server-islands.md

```diff
@@ -0,0 +1,147 @@
+
+- Component `name` as described in the naming algorithm.
+- `props` passed to the component. An island can be rendered multiple times; the props are representative of a particular usage.
+- `slots` that are passed to the island component.
+
+## Hydration
+
+The hydration script performs the following steps:
+
+- Creates an HTTP request to `/_server-islands/[name]`
+- Consumes the body of the request into a string.
+- Turns the string of HTML into a document fragment.
+- Removes the fallback content, if there is any.
+- Injects the new fragment.
+- Removes the script.
+
+# Testing Strategy
+
+This feature spans multiple parts of Astro so it will be tested in layers:
+
+## Compiler
+
+- The compiler piece of this mostly deals with the metadata that is returned. So integration/wasm tests will be added to verify the right output.
+
+## Rendering
+
+- Fixture tests for server generated content, dev and build, to ensure the script is emitted for islands.
+
+## E2E
+
+- Playwright tests to verify the island hydrates, requests the server contents, and renders it properly on the client. 
+
+# Drawbacks
+
+- There is overlap between this feature and client directives, particularly `client:only` which can include fallback content. It is hard to explain why you would use server islands over this feature. One reason is that client directives that fetch from an API cause a waterfall that is not included with server directives which only have a small inlined script.
+
+# Alternatives
+
+The major alternative implementation idea is to not fetch the islands via a script but rather to do so inside of an Edge function and then stitch together the HTML as the islands stream in. Such an approach would have these advantages:
+
+- Prevention of a waterfall caused by the island only being fetched once the page loads.
+- No fallback content needed. Fallback requires design considerations, using an Edge function would be more akin to SSR.
+
+However this approach has some downsides:
+
+- Eliminates the caching advantage gained by the script approach. Since the edge function injects personalized content the page cannot be cached globally.
+- Only works will with Edge functions, so limited choice of hosts. Loss of portability.
+- The page's main content will often be delayed from being visible to the user as it is blocked by server islands being fetched further up in the page.
+
+# Adoption strategy
+
+- Experimental release while the stage 3 RFC goes through revisions.
+- This is an opt-in feature that does not include any breaking changes to existing features. Only users who want to use it will.
+- This feature requires compiler integration so there are no similar features in the Astro ecosystem.
+
+# Unresolved Questions
+
+- During stage 2 there was some discussion about `props` which get serialized to the island. It could make sense to encrypt them to prevent mistakening leaking secrets. How/if this can be done hasn't been determined yet.
```

Comment:
An alternative is signing the props, which would address the potential issue of untrusted props - though not the one of leaking secrets.

Reply:
I think it would be best to use a built-time-generated symmetric key use with some algorithm available under the Web Crypto standard. That way it is available on basically every JS server runtime.